### PR TITLE
Add harnesses for all public functions of `Layout`

### PR DESCRIPTION
**Pull request by tautschnig**
Exercise all public member functions of `Layout` and assert properties over the result. Some of those should, perhaps, become `ensures` clauses.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/alloc/layout.rs [hunk 1]
```diff
@@ -514,4 +542,190 @@ mod verify {
             assert_eq!(layout.align(), a);
         }
     }
+
+    // pub const fn size(&self) -> usize
+    #[kani::proof]
```
Comment (celinval) [inline library/core/src/alloc/layout.rs:547]: Why not `proof_for_contract`?

File: library/core/src/alloc/layout.rs [hunk 2]
```diff
@@ -503,6 +515,22 @@ impl fmt::Display for LayoutError {
 mod verify {
     use super::*;
 
+    // pub const fn from_size_align(size: usize, align: usize) -> Result<Self, LayoutError>
+    #[kani::proof_for_contract(Layout::from_size_align)]
+    pub fn check_from_size_align() {
+        let s = kani::any::<usize>();
+        let a = kani::any::<usize>();
+
+        if let Ok(layout) = Layout::from_size_align(s, a) {
```
Comment (celinval) [inline library/core/src/alloc/layout.rs:524]: Can you implement `kani::Arbitrary` and `kani::Invariant` for Layout instead?

File: library/core/src/alloc/layout.rs [hunk 3]
```diff
@@ -514,4 +542,190 @@ mod verify {
...
+        let array : [i32; 2] = [1, 2];
+        let layout = Layout::for_value::<[i32]>(&array[1 .. 1]);
+        assert!(layout.align().is_power_of_two());
+    }
+
+    // pub const unsafe fn for_value_raw<T: ?Sized>(t: *const T) -> Self
+    #[kani::proof_for_contract(Layout::for_value_raw)]
+    pub fn check_for_value_raw_i32() {
+        unsafe {
+            let layout = Layout::for_value_raw::<[i32]>(&[] as *const [i32]);
```
Comment (celinval) [inline library/core/src/alloc/layout.rs:590]: For unsafe functions, it would be great if we could cover different safety scenarios.

**Review by feliperodri — COMMENTED** [review]
Some comments from https://github.com/model-checking/verify-rust-std/pull/42 also apply here. So I think maybe we should merge that previous PR first and implement some of the cleanups requested there in this one (e.g., using `kani::Arbitrary` and moving assertions to ensures clauses).

File: library/core/src/alloc/layout.rs (hunk 3 above)
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:590]: Hmm, yes, but how many would we consider "good enough?" And for this one specifically I believe we are lacking the means to express the different cases as `requires` clauses that the function's documentation calls out.

File: library/core/src/alloc/layout.rs (hunk 2 above)
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:524]: See https://github.com/model-checking/verify-rust-std/pull/42#discussion_r1703916363

File: library/core/src/alloc/layout.rs (hunk 1 above)
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:547]: Same problem as https://github.com/model-checking/verify-rust-std/pull/42#discussion_r1703939852

File: library/core/src/alloc/layout.rs (hunk 3 above)
Comment (celinval) [reply library/core/src/alloc/layout.rs]: I don't have a number for you, but I would consider in this case adding a few cases with types that have different layouts

Comment (tautschnig) [reply library/core/src/alloc/layout.rs]: I have added another one, but need some help in producing objects for further two cases. TODOs added.

File: library/core/src/alloc/layout.rs [hunk 4]
```diff
@@ -324,6 +345,13 @@ impl Layout {
     /// On arithmetic overflow, returns `LayoutError`.
     #[unstable(feature = "alloc_layout_extra", issue = "55724")]
     #[inline]
+    // the below modulo operation might be too costly to prove
+    // #[ensures(|result| result.is_err() || n == 0 || result.as_ref().unwrap().0.size() % n == 0)]
+    #[ensures(|result| result.is_err() || n == 0 || result.as_ref().unwrap().0.size() >= self.size())]
```
Comment (celinval) [inline library/core/src/alloc/layout.rs:350]:
```suggestion
    #[cfg_attr(not(kani), ensures(|result| result.is_err() || n == 0 || result.as_ref().unwrap().0.size() % n == 0))]
    #[cfg_attr(kani, ensures(|result| result.is_err() || n == 0 || result.as_ref().unwrap().0.size() >= self.size()))]
```

**Review by celinval — COMMENTED** [review]
I haven't finished yet

File: library/core/src/alloc/layout.rs [hunk 5]
```diff
@@ -167,6 +177,8 @@ impl Layout {
     #[rustc_const_unstable(feature = "const_alloc_layout", issue = "67521")]
     #[must_use]
     #[inline]
+    #[requires(mem::align_of_val(t).is_power_of_two())]
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:180]: I don't agree this is redundant: this is a sanity check on what rustc produces.

File: library/core/src/alloc/layout.rs [hunk 6]
```diff
@@ -203,6 +215,7 @@ impl Layout {
     #[unstable(feature = "layout_for_ptr", issue = "69835")]
     #[rustc_const_unstable(feature = "const_alloc_layout", issue = "67521")]
     #[must_use]
+    #[ensures(|result| result.align().is_power_of_two())]
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:218]: How would that interact with contracts? Isn't https://model-checking.github.io/kani/crates/doc/kani/invariant/trait.Invariant.html specific to Kani, so would not be of any use for verification approaches (unlike the above `ensures` clause, which is meant to eventually become standard)?

File: library/core/src/alloc/layout.rs [hunk 7]
```diff
@@ -152,6 +159,9 @@ impl Layout {
     #[rustc_const_stable(feature = "alloc_layout_const_new", since = "1.42.0")]
     #[must_use]
     #[inline]
+    #[requires(mem::align_of::<T>().is_power_of_two())]
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:162]: Perhaps, but it's about the only thing we can add as contract to this function.

File: library/core/src/alloc/layout.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (celinval) [reply library/core/src/alloc/layout.rs:221]: We should add an interface like we did with ensures

File: library/core/src/alloc/layout.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (celinval) [reply library/core/src/alloc/layout.rs:162]: This function has no requirement, and it should always succeed. So I think we should remove the `requires`, which is the same as writing `#[requires(true)]`

File: library/core/src/alloc/layout.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (celinval) [reply library/core/src/alloc/layout.rs:180]: I don't think contracts is meant to check if the compiler is working properly.

File: library/core/src/alloc/layout.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (celinval) [reply library/core/src/alloc/layout.rs:221]: for now, you could use `#[cfg_attr(kani, ensures(|result| result.is_valid()))]`

**Comment by rahulku** [comment]
is this blocked on something?

File: library/core/src/alloc/layout.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tautschnig) [reply library/core/src/alloc/layout.rs]: I tried to use `Invariant` per the documentation, but got:
```
error[E0404]: expected trait, found derive macro `kani::Invariant`
   --> /verify-rust-std.git/library/core/src/alloc/layout.rs:583:10
    |
583 |     impl kani::Invariant for Layout {
    |          ^^^^^^^^^^^^^^^ not a trait

```

[the remark this answers is not in the document]
Comment (tautschnig) [reply library/core/src/alloc/layout.rs]: It seems that `Invariant` is not in `kani_core`.

[the remark this answers is not in the document]
Comment (celinval) [reply library/core/src/alloc/layout.rs]: @jaisnan

**Review by celinval — APPROVED** [review]
Sorry, I was hoping the other stuff would be in by now. So let's unblock this PR and we can go back and improve this once the Invariant PR merges.

**Review by carolynzech — APPROVED** [review]

**Pull request merged**